R/add-subvariable.R

```diff
@@ -22,15 +23,59 @@ addSubvariable <- function (variable, subvariable) {
     crPATCH(shojiURL(variable, "catalogs", "subvariables"),
         body=toJSON(sapply(new.urls, emptyObject, simplify=FALSE)))
 
-    ## Workaround because apparently bind/rebind isn't retaining the order
+        ## Workaround because apparently bind/rebind isn't retaining the order
     crPATCH(self(variable),
         body=toJSON(list(subvariables=I(c(subvar.urls, new.urls)))))
 
     ## Refresh and return
     dropCache(datasetReference(variable))
-    invisible(refresh(variable))
+    return(invisible(refresh(variable)))
 }
 
 #' @rdname addSubvariable
 #' @export
 addSubvariables <- addSubvariable
+
+addSubvarDef <- function (var, subvar) {
+    ## Input can be a variable, subvariable, dataset subset or
+    ## a mixed or uniform list of variables and subvariables this
+    ## wraps single entries in a list for type consistency.
+    if (inherits(subvar, "VariableDefinition") ||
+            is.variable(subvar)) {
+        ## wrap single variables in list
+        subvar <- list(subvar)
+    }
+
+    vardefs <- vapply(subvar,
+        function(x) inherits(x, "VariableDefinition"),
+        logical(1))
+
+    out <- vector("list", length(subvar))
+
+    if (any(vardefs)) {
+        ds <- loadDataset(datasetReference(var))
+        var_cat_url <- shojiURL(ds, "catalogs", "variables")
```

Comment:
Perhaps a better way to do this, since I see that this is largely copy-and-pasted from `addVariables`, would be to factor https://github.com/Crunch-io/rcrunch/blob/master/R/add-variable.R#L37-L55 out into a function and reuse it here. That way, you'd also get test coverage for the error handling code below for free, which this branch doesn't have.